File: js/input.js
const { DateTime } = require("luxon"); // Sử dụng Luxon để xử lý ngày giờ

function parseDate(dateStr, hours) {
    const dateFormat = "dd-MM-yyyy";
    const originalDate = DateTime.fromFormat(dateStr, dateFormat);
    return originalDate.plus({ hours });
}

function findWorker(workers, workerId) {
    const worker = workers.find((worker) => worker._id === workerId);
    return worker ? worker.id : null;
}

function input(commands, workers) {
    const newCommands = [];
    const inprogressCommands = [];
    let newCommandCount = 1;
    let numOperations = 0;
    const commandMap = {};
    const machineMap = {};
    const workerMap = {};

    // Xử lý commands
    for (const command of commands) {
        if (command.status === "new") {
            command.id = newCommandCount;
            newCommands.push(command);
            newCommandCount++;
            numOperations += command.operations.length;
            commandMap[command.id] = command.command_id;
        } else if (command.status === "inprogress") {
            inprogressCommands.push(command);
        }
    }

    // Xử lý workers
    let workerCount = 1;
    for (const worker of workers) {
        worker.id = workerCount;
        workerMap[workerCount] = worker.worker_id;
        workerCount++;
    }

    // Xử lý machines
    let machineCount = 1;
    const machines = {};
    for (const command of newCommands) {
        for (const operation of command.operations) {
            for (const resource of operation.resources) {
                const machineId = resource.machine_id;
                if (!machines[machineId]) {
                    machines[machineId] = machineCount;
                    machineMap[machineCount] = machineId;
                    machineCount++;
                }
            }
        }
    }

    const keyMap = {
        command_map: commandMap,
        machine_map: machineMap,
        worker_map: workerMap,
    };

    const numJobs = newCommands.length;
    const numMachines = Object.keys(machines).length;
    const numWorkers = workers.length;
    const numOperationsPerJob = newCommands.map((job) => job.operations.length);

    const lotSizes = Array(numJobs).fill(0);
    const startDates = Array(numJobs).fill("");
    const hprs = Array.from({ length: numOperations }, () =>
        Array(numMachines).fill(0)
    );
    const availMachines = Array.from({ length: numOperations }, () => []);
    const availWorkers = Array.from({ length: numOperations }, () =>
        Array.from({ length: numMachines }, () => [])
    );

    const maxOperationsPerJob = Math.max(...numOperationsPerJob);
    const jobOperations = Array.from({ length: numJobs + 1 }, () =>
        Array(maxOperationsPerJob * numMachines + 1).fill(-1)
    );
    const jobTotalOperations = Array.from({ length: numJobs + 1 }, () =>
        Array(numOperations * numMachines + 1).fill(-1)
    );
    const jobLots = Array.from({ length: numJobs + 1 }, () =>
        Array(maxOperationsPerJob * numMachines + 1).fill(-1)
    );
    const jobOperationIndex = Array.from({ length: numJobs + 1 }, () =>
        Array(numOperations).fill(-1)
    );
    const jobMainOperation = Array.from({ length: numJobs + 1 }, () =>
        Array(maxOperationsPerJob + 1).fill(-1)
    );
    const prevOperations = Array.from({ length: numJobs + 1 }, () =>
        Array(maxOperationsPerJob * numMachines + 1).fill(-1)
    );

    const machinesAvailStartTime = Array.from(
        { length: numMachines },
        () => []
    );
    const machinesAvailEndTime = Array.from({ length: numMachines }, () => []);
    const workersAvailStartTime = Array.from({ length: numWorkers }, () => []);
    const workersAvailEndTime = Array.from({ length: numWorkers }, () => []);

    let opIndex = 0;
    let totalOperationCount = 0;

    for (const job of newCommands) {
        const jobId = job.id;
        const lotSize = job.quantity;
        const startDate = job.start_date;
        lotSizes[jobId - 1] = lotSize;
        startDates[jobId - 1] = startDate;

        let opInJobCount = 1;
        let mainOperationInJob = 1;

        for (const operation of job.operations) {
            const preOperation = operation.prevOperation;
            const prevOperationId =
                Array.isArray(preOperation) && preOperation.length > 0
                    ? preOperation[0]
                    : preOperation;

            jobOperationIndex[jobId][opIndex] = mainOperationInJob;
            jobMainOperation[jobId][mainOperationInJob] = opIndex;

            for (const resource of operation.resources) {
                const machineId = machines[resource.machine_id];
                const minExp = resource.minExp;
                const role = resource.role;

                const availWorker = workers
                    .filter(
                        (worker) => worker.role === role && worker.exp >= minExp
                    )
                    .map((worker) => worker.id);

                const lotIndex = machineId - 1;
                hprs[opIndex][machineId - 1] = resource.hprs;
                availMachines[opIndex].push(machineId);
                availWorkers[opIndex][machineId - 1] = availWorker;

                jobOperations[jobId][opInJobCount] = opIndex;
                jobTotalOperations[jobId][opInJobCount] = totalOperationCount;
                jobLots[jobId][opInJobCount] = lotIndex;

                if (
                    prevOperationId !== null &&
                    jobMainOperation[jobId][prevOperationId] !== undefined
                ) {
                    prevOperations[jobId][opInJobCount] =
                        jobMainOperation[jobId][prevOperationId];
                } else {
                    prevOperations[jobId][opInJobCount] = -1; // Giá trị mặc định nếu không hợp lệ
                }

                opInJobCount++;
                totalOperationCount++;
            }

            opIndex++;
            mainOperationInJob++;
        }
    }

    const parsedDates = startDates.map((date) =>
        DateTime.fromFormat(date, "dd-MM-yyyy")
    );
    const minDate = parsedDates.reduce((min, date) =>
        date < min ? date : min
    );
    const minStartDateStr = minDate.toFormat("dd-MM-yyyy");
    const startDateInHours = parsedDates.map(
        (date) => date.diff(minDate, "hours").hours
    );

    for (const command of inprogressCommands) {
        for (const operation of command.operations) {
            for (const resource of operation.resources) {
                const machineId = machines[resource.machine_id];
                const workerId = findWorker(workers, resource.worker_id);

                const startTime = parseDate(
                    resource.start_date,
                    resource.start_hours
                );
                const endTime = parseDate(
                    resource.end_date,
                    resource.end_hours
                );

                if (startTime > minDate) {
                    machinesAvailStartTime[machineId - 1].push(
                        startTime.diff(minDate, "hours").hours
                    );
                    if (workerId !== null) {
                        workersAvailStartTime[workerId - 1].push(
                            startTime.diff(minDate, "hours").hours
                        );
                    }
                }

                if (endTime > minDate) {
                    machinesAvailEndTime[machineId - 1].push(
                        endTime.diff(minDate, "hours").hours
                    );
                    if (workerId !== null) {
                        workersAvailEndTime[workerId - 1].push(
                            endTime.diff(minDate, "hours").hours
                        );
                    }
                }
            }
        }
    }

    return {
        keyMap,
        numMachines,
        numWorkers,
        numOperations,
        numJobs,
        numOperationsPerJob,
        minStartDateStr,
        startDateInHours,
        availMachines,
        availWorkers,
        lotSizes,
        jobOperations,
        jobLots,
        jobTotalOperations,
        jobOperationIndex,
        prevOperations,
        hprs,
        machinesAvailStartTime,
        machinesAvailEndTime,
        workersAvailStartTime,
        workersAvailEndTime,
    };
}

module.exports = { input };

File: python/input.py
import numpy as np
from datetime import datetime, timedelta
from command_input import CommandInput
from worker_input import WorkerInput

def parse_date(date_str, hours):
    date_format = "%d-%m-%Y"

    original_date = datetime.strptime(date_str, date_format)

    new_date = original_date + timedelta(hours=hours)

    return new_date


def find_worker(workers, worker_id):
    worker = next((worker for worker in workers if worker["_id"] == worker_id), None)

    if worker == None:
        return None
    else:
        return worker["id"]


def input(commands: list[CommandInput], workers: list[WorkerInput]):
    new_commands = []
    inprogress_commands = []
    new_command_count = 1
    num_operations = 0
    command_map = {}
    machine_map = {}
    worker_map = {}

    for command in commands:
        if command.status == "new":
            command.id = new_command_count
            new_commands.append(command)
            new_command_count += 1
            num_operations += len(command.operations)
            command_map[command.id] = command.command_id

        elif command.status == "inprogress":
            inprogress_commands.append(command)

    worker_count = 1
    for worker in workers:
        worker.id = worker_count
        worker_map[worker_count] = worker.worker_id
        worker_count += 1

    machine_count = 1
    machines = {}
    for command in new_commands:
        for operations in command.operations:
            for resource in operations.resources:
                machine_id = resource.machine_id
                if machine_id not in machines:
                    machines[machine_id] = machine_count
                    machine_map[machine_count] = machine_id
                    machine_count += 1

    key_map = {
        "command_map": command_map,
        "machine_map": machine_map,
        "worker_map": worker_map,
    }

    num_jobs = len(new_commands)
    num_machines = len(machines)
    num_workers = len(workers)
    num_operations_per_job = []
    for job in new_commands:
        num_operations_per_job.append(len(job.operations))

    lot_sizes = np.zeros(num_jobs)
    start_dates = ["" for _ in range(num_jobs)]
    hprs = np.zeros((num_operations, num_machines))
    avail_machines = [[] for _ in range(num_operations)]
    avail_workers = [[[] for _ in range(num_machines)] for _ in range(num_operations)]
    max_operations_per_job = max(num_operations_per_job)
    job_operations = np.ones(
        (num_jobs + 1, max_operations_per_job * num_machines + 1), dtype=int
    ) * (-1)
    job_total_operations = np.ones(
        (num_jobs + 1, num_operations * num_machines + 1), dtype=int
    ) * (-1)
    job_lots = np.ones(
        (num_jobs + 1, max_operations_per_job * num_machines + 1), dtype=int
    ) * (-1)
    job_operation_index = np.ones((num_jobs + 1, num_operations), dtype=int) * (-1)
    job_main_operation = np.ones(
        (num_jobs + 1, max_operations_per_job + 1), dtype=int
    ) * (-1)
    prev_operations = np.ones(
        (num_jobs + 1, max_operations_per_job * num_machines + 1), dtype=int
    ) * (-1)
    machines_avail_start_time = [[] for _ in range(num_machines)]
    machines_avail_end_time = [[] for _ in range(num_machines)]
    workers_avail_start_time = [[] for _ in range(num_workers)]
    workers_avail_end_time = [[] for _ in range(num_workers)]

    jobs = new_commands

    op_index = 0
    total_operation_count = 0
    for job in jobs:
        job_id = job.id
        lot_size = job.quantity
        start_date = job.start_date
        lot_sizes[job_id - 1] = lot_size
        start_dates[job_id - 1] = start_date
        op_in_job_count = 1
        main_operation_in_job = 1
        for operation in job.operations:
            preOperation = operation.prevOperation
            prev_operation_id = None

            if isinstance(preOperation, int):
                prev_operation_id = preOperation
            elif isinstance(preOperation, list) and len(preOperation) > 0:
                prev_operation_id = preOperation[0]

            job_operation_index[job_id, op_index] = main_operation_in_job
            job_main_operation[job_id, main_operation_in_job] = op_index

            for resource in operation.resources:
                machine_id = machines[resource.machine_id]
                minExp = resource.minExp
                role = resource.role
                avail_worker = [
                    worker.id
                    for worker in workers
                    if worker.role == role and worker.exp >= minExp
                ]
                lot_index = machine_id - 1
                hprs[op_index, machine_id - 1] = resource.hprs
                avail_machines[op_index].append(machine_id)
                avail_workers[op_index][machine_id - 1] = avail_worker

                job_operations[job_id, op_in_job_count] = op_index
                job_total_operations[job_id, op_in_job_count] = total_operation_count
                job_lots[job_id, op_in_job_count] = lot_index

                if prev_operation_id != None:
                    prev_operations[job_id, op_in_job_count] = job_main_operation[
                        job_id, prev_operation_id
                    ]

                op_in_job_count += 1
                total_operation_count += 1

            op_index += 1
            main_operation_in_job += 1

    parsed_dates = [datetime.strptime(st, "%d-%m-%Y") for st in start_dates]
    min_date = min(parsed_dates)
    min_start_date_str = min_date.strftime("%d-%m-%Y")
    start_date_in_hours = [
        (date - min_date).total_seconds() / 3600 for date in parsed_dates
    ]

    for command in inprogress_commands:
        for operations in command.operations:
            for resource in operations.resources:
                _machine_id = resource.machine_id
                if _machine_id in machines:
                    start_time = parse_date(resource.start_date, resource.start_hours)
                    end_time = parse_date(resource.end_date, resource.end_hours)
                    machine_id = machines[_machine_id]
                    worker_id = find_worker(workers, resource.worker_id)

                    if start_time > min_date:
                        machines_avail_start_time[machine_id - 1].append(
                            (start_time - min_date).total_seconds() / 3600
                        )
                        if worker_id != None:
                            workers_avail_start_time[worker_id - 1].append(
                                (start_time - min_date).total_seconds() / 3600
                            )

                    if end_time > min_date:
                        machines_avail_end_time[machine_id - 1].append(
                            (end_time - min_date).total_seconds() / 3600
                        )
                        if worker_id != None:
                            workers_avail_end_time[worker_id - 1].append(
                                (end_time - min_date).total_seconds() / 3600
                            )

    print(prev_operations)
    return (
        key_map,
        num_machines,
        num_workers,
        num_operations,
        num_jobs,
        num_operations_per_job,
        min_start_date_str,
        start_date_in_hours,
        avail_machines,
        avail_workers,
        lot_sizes,
        job_operations,
        job_lots,
        job_total_operations,
        job_operation_index,
        prev_operations,
        hprs,
        machines_avail_start_time,
        machines_avail_end_time,
        workers_avail_start_time,
        workers_avail_end_time,
    )
